src/string.rs

```diff
@@ -173,40 +188,183 @@ impl<'de> serde::Deserialize<'de> for IString{
+                test_one!($macro!, $frame1!($a), $frame2!($b));
+                test_one!($macro!, $frame2!($a), $frame1!($b));
+            };
+            ($macro:tt!, $frame2:tt!, $a:literal, $b:literal) => {
+                test_all_frame_combos!($macro!, frame_i_rc!, $frame2!, $a, $b);
+                test_all_frame_combos!($macro!, frame_i_static!, $frame2!, $a, $b);
+            };
+            ($macro:tt!, $a:literal, $b:literal) => {
+                test_all_frame_combos!($macro!, frame_deref!, $a, $b);
+                test_all_frame_combos!($macro!, frame_noop!, $a, $b);
+                test_all_frame_combos!($macro!, frame_string!, $a, $b);
+                test_all_frame_combos!($macro!, frame_string_ref!, $a, $b);
+            };
+        }
 
-        assert_ne!(IString::Static("foo"), IString::Static("bar"));
-        assert_ne!(IString::Static("foo"), IString::Rc(Rc::from("bar")));
-        assert_ne!(IString::Rc(Rc::from("foo")), IString::Rc(Rc::from("bar")));
+        test_all_frame_combos!(assert_eq!, "foo", "foo");
+        test_all_frame_combos!(assert_ne!, "foo", "bar");
     }
 
     #[test]
-    fn string_cmp() {
-        assert_eq!(IString::from("foo"), "foo");
-        assert_eq!(IString::from("foo"), String::from("foo"));
-        assert_eq!(IString::from("foo"), &String::from("foo"));
+    fn partial_cmp_strings() {
+        macro_rules! test_one {
+            ($res:expr, $a:expr, $b:expr) => {
+                assert_eq!(Some($res), PartialOrd::partial_cmp(&$a, &$b));
+            };
+        }
+
+        macro_rules! test_all_frame_combos {
+            ($res:expr, $frame1:tt!, $frame2:tt!, $a:literal, $b:literal) => {
+                // 12, 21 - i_rc-deref, deref-i_rc, ..., static-string_ref, string_ref-static
+                test_one!($res, $frame1!($a), $frame2!($b));
+                test_one!($res, $frame2!($a), $frame1!($b));
+            };
+            ($res:expr, $frame2:tt!, $a:literal, $b:literal) => {
+                test_all_frame_combos!($res, frame_i_rc!, $frame2!, $a, $b);
+                test_all_frame_combos!($res, frame_i_static!, $frame2!, $a, $b);
+            };
+            ($res:expr, $a:literal, $b:literal) => {
+                test_all_frame_combos!($res, frame_deref!, $a, $b);
+                test_all_frame_combos!($res, frame_noop!, $a, $b);
+                test_all_frame_combos!($res, frame_string!, $a, $b);
+                test_all_frame_combos!($res, frame_string_ref!, $a, $b);
+            };
+        }
+
+        test_all_frame_combos!(Ordering::Equal, "foo", "foo");
+        test_all_frame_combos!(Ordering::Greater, "foo", "bar");
+        test_all_frame_combos!(Ordering::Less, "bar", "foo");
+        test_all_frame_combos!(Ordering::Greater, "foobar", "foo");
+        test_all_frame_combos!(Ordering::Less, "foo", "foobar");
     }
 
     #[test]
-    fn static_string() {
+    fn const_string() {
```

Comment:
I think it was called "static" because it works on the type `&'static str`. But yeah it's a const xD I don't know what's best

Reply:
My thought was that the test is that `IString` may be used in a `const` context. Haven't checked if `Rc` (struct and the `IString` variant) may be used in a `const` context, but this test confirms that `IString`, at least in some way, can.